### PR TITLE
fix: do not crash on error in `fs.walk` filter

### LINKED ISSUE #18250: Bug: unhandled error in `lintFiles` with glob patterns when config is invalid
### Environment

Node version: v21.7.1
npm version: v10.5.0
Local ESLint version: v9.0.0-rc.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 23.4.0

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
export default {
  invalid: 'foobar',
};
```
</details>

```js
// index.js
import { ESLint } from 'eslint';

const eslint = new ESLint();
try {
  await eslint.lintFiles('*.js');
} catch {
  console.log('An error should be caught because the config is invalid');
}
console.log('This message should be printed');
```

### What did you expect to happen?

`lintFiles` should throw an error in the current execution context because the config is invalid.

### What actually happened?

The process terminated abnormally while running `lintFiles` because of an uncaught error in an asynchronous callback.

```console
$ node ./index.js
.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:270
                throw new UnexpectedKeyError(key);
                ^

UnexpectedKeyError: Unexpected key "invalid" found.
    at ObjectSchema.validate (.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:270:23)
    at .../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:239:18
    at Array.reduce (<anonymous>)
    at ObjectSchema.merge (.../project/node_modules/@humanwhocodes/object-schema/src/object-schema.js:237:24)
    at .../project/node_modules/@humanwhocodes/config-array/api.js:935:42
    at Array.reduce (<anonymous>)
    at FlatConfigArray.getConfig (.../project/node_modules/@humanwhocodes/config-array/api.js:934:39)
    at FlatConfigArray.isFileIgnored (.../project/node_modules/@humanwhocodes/config-array/api.js:962:15)
    at .../project/node_modules/eslint/lib/eslint/eslint-helpers.js:322:49
    at Array.reduce (<anonymous>)
```

The callback in question is the [`entryFilter` of `@nodelib/fs.walk`](https://github.com/eslint/eslint/blob/de408743b5c3fc25ebd7ef5fb11ab49ab4d06c36/lib/eslint/eslint-helpers.js#L291-L331), which apparently is not expected to throw errors.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/stackblitz-starters-swoxsq

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by fasttime**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

Fixes #18250

#### What changes did you make? (Give an overview)

I replaced [`util.promisify`](https://github.com/eslint/eslint/compare/issue-18250?expand=1#diff-87d53094b12d82e4c11a0e1167d79cf2f471d2f5e5ebb6fc483e891f9dc87a5aL27) around `fs.walk` with a plain promise. In this way, the promise can be rejected not only when the implicit callback provided by `util.promisify` reports an error, but also when an error is caught in one of the filter functions [`deepFilter`](https://github.com/eslint/eslint/compare/issue-18250?expand=1#diff-87d53094b12d82e4c11a0e1167d79cf2f471d2f5e5ebb6fc483e891f9dc87a5aL285-L290) or [`entryFilter`](https://github.com/eslint/eslint/compare/issue-18250?expand=1#diff-87d53094b12d82e4c11a0e1167d79cf2f471d2f5e5ebb6fc483e891f9dc87a5aL291-L297). This allows Node.js to catch the rejection in the current execution context (where the promise was created) and avoid an unhandled exception.

Related issue in NodeLib: https://github.com/nodelib/nodelib/issues/105

### CODE REVIEW

File: tests/lib/eslint/eslint.js
Comment (fasttime) [inline tests/lib/eslint/eslint.js:4515]: This test adds code coverage for [this line](https://github.com/eslint/eslint/pull/18295/files#diff-87d53094b12d82e4c11a0e1167d79cf2f471d2f5e5ebb6fc483e891f9dc87a5aR366), for the case that the callback function is called with an error.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Would like @nzakas to verify.

**Review by nzakas — COMMENTED** [review]
Left one comment about the code but overall LGTM.

I'd like to wait and see what kind of response we get to https://github.com/nodelib/nodelib/issues/105 before merging, though, as I think getting them to fix this behavior is the better approach.

File: lib/eslint/eslint-helpers.js
Comment (fasttime) [reply lib/eslint/eslint-helpers.js:369]: If there are any listeners for `multipleResolves` events on the process and a promise is settled twice, a deprecation warning will be printed to stderr (the warning is emitted when a listener is first called, not when listeners are added).

If this is not a concern, we could remove the check and save a couple of lines.

Comment (nzakas) [reply lib/eslint/eslint-helpers.js:369]: Given how long ago this was deprecated, I think it's fair to say we don't need to care about this.

Comment (fasttime) [reply lib/eslint/eslint-helpers.js]: Done in 74e8e49c3e25db6a40b55acff0bba00aa5f5c7c5.

**Review by nzakas — APPROVED** [review]
LGTM. I'd still like to hold off until we know whether or not fs.walk will be fixed.

**Comment by fasttime** [comment]
I'll convert this to a draft so we don't merge it accidentally.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

If the problem isn't fixed in the dependencies by the release of ESLint v9.1.0 this week, I think we should merge this PR to fix the bug we have. Then, if and when the problem is fixed in the dependencies in the future, we could revert these code changes (but keep the tests anyway).

**Comment by nzakas** [comment]
Given that we still haven't heard back from nodelib, I think it's fine to merge this into this week's release.

**Comment by fasttime** [comment]
Merging now.

**Pull request merged**
